libmscore/realizedharmony.cpp

```diff
@@ -0,0 +1,555 @@
+                  QMapIterator<int, int> i(intervals);
+                  while (i.hasNext()) {
+                        i.next();
+                        notes.insert((rootPitch + (i.key() % 128)) % PITCH_DELTA_OCTAVE +
+                                      5*PITCH_DELTA_OCTAVE, i.value());
+                        }
+                  }
+                  break;
+            case Voicing::DROP_2:
+                  {
+                  //select 4 notes from list
+                  QMap<int, int> intervals = normalizeNoteMap(getIntervals(rootTpc, literal), rootTpc, rootPitch, 4);
+                  QMapIterator<int, int> i(intervals);
+                  i.toBack();
+
+                  int counter = 0; //counter to drop the second note
+                  while (i.hasPrevious()) {
+                        i.previous();
+                        if (++counter == 2)
+                              notes.insert(i.key() + 4*PITCH_DELTA_OCTAVE, i.value());
+                        else
+                              notes.insert(i.key() + 5*PITCH_DELTA_OCTAVE, i.value());
+                        }
+                  }
+                  break;
+            case Voicing::THREE_NOTE:
+                  {
+                  //Insert 2 notes in the octave above middle C
+                  QMap<int, int> intervals = normalizeNoteMap(getIntervals(rootTpc, literal), rootTpc, rootPitch, 2, true);
+                  QMapIterator<int, int> i(intervals);
+
+                  i.next();
+                  notes.insert(i.key() + 5*PITCH_DELTA_OCTAVE, i.value());
+
+                  i.next();
+                  notes.insert(i.key() + 5*PITCH_DELTA_OCTAVE, i.value());
+                  }
+                  break;
+            case Voicing::FOUR_NOTE:
+                  {
+                  //four note voicing, drop every other note
+                  QMap<int, int> relIntervals = getIntervals(rootTpc, literal);
+                  QMap<int, int> intervals = normalizeNoteMap(relIntervals, rootTpc, rootPitch, 3, true);
+                  QMapIterator<int, int> i(intervals);
+                  i.toBack();
+
+                  int counter = 0; //how many notes have been added
+                  while (i.hasPrevious()) {
+                        i.previous();
+
+                        if (counter % 2)
+                              notes.insert(i.key() + 4*PITCH_DELTA_OCTAVE, i.value());
+                        else
+                              notes.insert(i.key() + 5*PITCH_DELTA_OCTAVE, i.value());
+                        ++counter;
+                        }
+                  }
+                  break;
+            case Voicing::SIX_NOTE:
+                  {
+                  //six note voicing, drop every other note
+                  QMap<int, int> relIntervals = getIntervals(rootTpc, literal);
+                  QMap<int, int> intervals = normalizeNoteMap(relIntervals, rootTpc, rootPitch, 5, true);
```

Comment:
Looks like 3 vs 5 is the only difference between this code and the case above?  Could they be collapsed / refactored?

Reply:
Ah yes it seems so, I'll look into it